### PR TITLE
bug fix to make QCPath values out of 0 to 31 valid

### PR DESCRIPTION
**Pull request by ypwang19**
## Description
In the the VIIRS AOD Level 2 data file, QCPath is set with a valid range of 0-31. The values out of 0-31 are not read in properly by the converters, so that the retrieval surface is not correctly identified. This PR is to fix this minor bug in reading QCPath, so that values out of 0-31 are valid. 

## Dependencies

List the other PRs that this PR is dependent on:
- None.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have run the unit tests before creating the PR

### CODE REVIEW

File: src/compo/viirs_aod2ioda.py [hunk 1]
```diff
@@ -99,6 +99,7 @@ def _read(self):
             vals = ncd.variables['AOD550'][:].ravel()
             errs = ncd.variables['Residual'][:].ravel()
             qcpath = ncd.variables['QCPath'][:].ravel()
+            qcpath = np.where(qcpath>-1,qcpath,np.nan)
```
Comment (CoryMartin-NOAA) [inline src/compo/viirs_aod2ioda.py:102]:
```suggestion
            qcpath = np.where(qcpath>-1, qcpath, np.nan)
```
I hope this suggestion makes it pass coding norms, may need to adjust the greater than too

**Review by jeromebarre — CHANGES_REQUESTED** [review]
This PR replaces the negative values in QCPath to `np.nan`. I am not sure from which values QC path is supposed to range. Cannot really find the documentation of the product that late on a Friday night. Would be good to have a link to the product documentation in the converter description (see for example trace gas converters). For the negative values, they might mean something. If they are -999 or something like that, we would might not want to throw NaNs but we should filter out those values using the already existing mask in the converter. Thanks a lot!

**Comment by jeromebarre** [comment]
And the coding norms tests is still failing. Just run the first test (compo coding norms) with the -VV option. It specify where your syntax is failing the test. I usually do that right before or right after I issue a PR in iodaconv. Thanks again.

**Comment by ypwang19** [comment]
> This PR replaces the negative values in QCPath to `np.nan`. I am not sure from which values QC path is supposed to range. Cannot really find the documentation of the product that late on a Friday night. Would be good to have a link to the product documentation in the converter description (see for example trace gas converters). For the negative values, they might mean something. If they are -999 or something like that, we would might not want to throw NaNs but we should filter out those values using the already existing mask in the converter. Thanks a lot!


The variable in the data file has the following attributes:
```
byte QCPath(Rows, Columns) ;
                QCPath:long_name = "Flags for retrieval path (0-No/1-Yes): bit 0: retrieval over water; bit 1: over bright land; bit 2: over glint water; bit 3: retrieval with SW scheme over land; bit 4: retrieval with SWIR scheme over land; bit 5: retrieval over bright-land algorithm, bit 6: dark land retrieval over bright land" ;
                QCPath:coordinates = "Longitude Latitude" ;
                QCPath:_FillValue = -128b ;
                QCPath:valid_range = 0b, 31b ;
                QCPath:units = "1" ;
```
The QCpath is supposed to range 0-127. The attribute `QCPath:valid_range = 0b, 31b` in the data file is a mistake, based on the communication with Hongqing from NESDIS data product team, and based on their document: https://www.star.nesdis.noaa.gov/jpss/documents/ATBD/ATBD_EPS_Aerosol_AOD_v3.4.pdf

Would it be better to set values below 0 to its missing value -128 (right after the the read in line) like this:
` qcpath = np.where(qcpath >= 0.0, qcpath, ncd.variables['QCPath']._FillValue)
`
and let the converter do the further quality control job? This line is just making values out of the "0-31 valid range" be valid, instead of "--".
I will also add some description near this line in the converter.

**Comment by jeromebarre** [comment]
> > This PR replaces the negative values in QCPath to `np.nan`. I am not sure from which values QC path is supposed to range. Cannot really find the documentation of the product that late on a Friday night. Would be good to have a link to the product documentation in the converter description (see for example trace gas converters). For the negative values, they might mean something. If they are -999 or something like that, we would might not want to throw NaNs but we should filter out those values using the already existing mask in the converter. Thanks a lot!
> 
> The variable in the data file has the following attributes:
> 
> ```
> byte QCPath(Rows, Columns) ;
>                 QCPath:long_name = "Flags for retrieval path (0-No/1-Yes): bit 0: retrieval over water; bit 1: over bright land; bit 2: over glint water; bit 3: retrieval with SW scheme over land; bit 4: retrieval with SWIR scheme over land; bit 5: retrieval over bright-land algorithm, bit 6: dark land retrieval over bright land" ;
>                 QCPath:coordinates = "Longitude Latitude" ;
>                 QCPath:_FillValue = -128b ;
>                 QCPath:valid_range = 0b, 31b ;
>                 QCPath:units = "1" ;
> ```
> 
> The QCpath is supposed to range 0-127. The attribute `QCPath:valid_range = 0b, 31b` in the data file is a mistake, based on the communication with Hongqing from NESDIS data product team, and based on their document: https://www.star.nesdis.noaa.gov/jpss/documents/ATBD/ATBD_EPS_Aerosol_AOD_v3.4.pdf
> 
> Would it be better to set values below 0 to its missing value -128 (right after the the read in line) like this: `qcpath = np.where(qcpath >= 0.0, qcpath, ncd.variables['QCPath']._FillValue)` and let the converter do the further quality control job? This line is just making values out of the "0-31 valid range" be valid, instead of "--". I will also add some description near this line in the converter.

We can either use the mask or set it as a value that we know how to filter in the obsspace during a JEDI execution. If you choose to do the latter please provide a test/example in UFO. Thanks!

**Comment by ypwang19** [comment]
> 

OK, I will mask out values out of 0-127. The QCPath will not go to the observation file or go through JEDI execution. It is only used to identify surface to assign observation error.

**Comment by CoryMartin-NOAA** [comment]
So my understanding of the issue here was that QCPath was not being properly converted from a bitmap to an integer. The QCPath is only used in computing the observation errors, otherwise it is ignored and not passed to the output file. Thus this PR is only changing the assigned initial errors and not masking out any observations.

**Comment by jeromebarre** [comment]
Ok sorry, I got confused and went too fast on this PR review. So QCpath is used to define the error here and not stored in the ioda file. So this PR changes only the way the error is specified. Meaning more data points now fall under: https://github.com/JCSDA-internal/ioda-converters/blob/d58dce0e5ef429fd3e58cde42512de004f6a2017/src/compo/viirs_aod2ioda.py#L135 I presume.
One of my question still remains though. Don't we want to get rid of the datapoints that have a wrong QCPath?

**Comment by ypwang19** [comment]
> Ok sorry, I got confused and went too fast on this PR review. So QCpath is used to define the error here and not stored in the ioda file. So this PR changes only the way the error is specified. Meaning more data points now fall under:
> 
> https://github.com/JCSDA-internal/ioda-converters/blob/d58dce0e5ef429fd3e58cde42512de004f6a2017/src/compo/viirs_aod2ioda.py#L135
> 
> I presume.
> One of my question still remains though. Don't we want to get rid of the datapoints that have a wrong QCPath?

Yes, more data fall under dark land error. The QCPath data is fine, the issue is they put a wrong valid range attribute to this variable by mistake. We fix it in this PR.

**Comment by jeromebarre** [comment]
Thanks for clarifying. The PR description was unclear to me at first.

**Pull request merged**